torch/csrc/LazyTensor.cpp

```diff
@@ -0,0 +1,300 @@
+};
+
+static GraphCache graph_cache;
+
+std::pair<
+    std::shared_ptr<torch::jit::GraphExecutor>,
+    std::vector<const LazyTensorImpl*>>
+getGraph(const LazyTensorImpl* lt_) {
+  // Construct a straight line repr of the LT
+  //  a = ?.to_lazy()
+  //  b = ?.to_lazy()
+  //  c = a @ b
+  //  d = c * 2
+  // becomes
+  //  *[0], @[0], 2, a, b
+  // that is, all LTs with exprs are "ExprName[index]", constants are constant
+  // and LTs with tensor bindings are LT*
+  // We can theoretically do memory comparison on these
+
+  std::vector<Item> record;
+  std::vector<c10::IValue> frontier = lt_->inps_;
+  std::vector<const LazyTensorImpl*> inputs;
+  while (frontier.size()) {
+    std::vector<c10::IValue> new_frontier;
+    for (const auto& ival : frontier) {
+      if (ival.isTensor()) {
+        auto t = ival.toTensor();
+        if (t.is_lazy()) {
+          auto* lt = static_cast<LazyTensorImpl*>(t.unsafeGetTensorImpl());
+          if (lt->schema_) {
+            // make into Item, push to record
+            record.emplace_back(
+                std::make_pair(lt->schema_->name(), lt->index_));
+            // maybe add to new_frontier
+            new_frontier.reserve(new_frontier.size() + lt->inps_.size());
+            new_frontier.insert(
+                new_frontier.end(), lt->inps_.begin(), lt->inps_.end());
+            continue;
+          } else {
+            inputs.emplace_back(lt);
+          }
+        }
+      }
+      record.emplace_back(ival);
+    }
+    frontier = new_frontier;
+  }
+
+  auto ge = graph_cache.get(record);
+  if (ge) {
+    return std::make_pair(ge, inputs);
+  } else {
+    inputs.clear();
+  }
+
+  auto g = std::make_shared<torch::jit::Graph>();
+  torch::jit::Value* output;
+  std::tie(output, inputs) = getValue(lt_, g);
+  g->registerOutput(output);
+  ge = std::make_shared<torch::jit::GraphExecutor>(g);
+  graph_cache.put(std::move(record), ge);
+
+  return std::make_pair(ge, inputs);
+}
+
+Tensor LazyTensorImpl::to_eager() const {
+  if (!schema_) {
+    auto t = inps_[0].toTensor();
+    return t;
+  }
+
+  std::shared_ptr<torch::jit::GraphExecutor> ge;
+  std::vector<const LazyTensorImpl*> inputs;
+  std::tie(ge, inputs) = getGraph(this);
+  torch::jit::Stack stack;
+  for (auto& inp : inputs) {
+    stack.push_back(inp->inps_[0]);
+  }
+
+  ge->run(stack);
+
+  // We registered only the output we care about
+  auto output_t = stack[0].toTensor();
+  if (output_t.is_variable()) {
+    return output_t;
+  }
+  auto output_var = torch::autograd::make_variable(output_t);
+  output_t = output_var.tensor_data();
+  TORCH_INTERNAL_ASSERT(output_t.is_variable());
+  return output_t;
+}
+
+} // namespace at
+
+namespace at {
+
+using namespace torch::jit;
+void lazyRun(const OperatorHandle& op, torch::jit::Stack* s) {
+  const FunctionSchema& fs = op.schema();
+  if (fs.name() == "aten::to_eager") {
+    auto t = pop(*s).toTensor();
+    auto lt = static_cast<LazyTensorImpl*>(t.unsafeGetTensorImpl());
```

Comment:
Ah! I think your problem is that you didn't register the native implementation of `to_eager` under LazyTensorTypeId. If you did, it would take precedence, and then this if-block would be dead. Having to_eager get a specialized implementation would be clearer structure, IMO.